### PR TITLE
Add the hold ptr type to mark the ownership.

### PR DESCRIPTION
**Pull request by Shylock-Hg**

### CODE REVIEW

File: src/common/base/HoldPtr.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    static_assert(!std::is_reference_v<T>);
+    explicit HoldPtr<T>(T* p) : ptr_{p} {}
+    HoldPtr<T>(const HoldPtr<T> &p) : ptr_(p.ptr_) {}
+    HoldPtr<T>(HoldPtr<T> &&p) : ptr_(p.ptr_) {
+        p.ptr_ = nullptr;
+    }
+    ~HoldPtr<T>() = default;
+
+    // Take care the ownership
+    // Don't own the retrun raw pointer
```
Comment (critical27) [inline src/common/base/HoldPtr.h:40]: typo

File: src/common/base/HoldPtr.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+template <typename T>
+struct HoldPtr {
+public:
+    static_assert(!std::is_reference_v<T>);
+    explicit HoldPtr<T>(T* p) : ptr_{p} {}
+    HoldPtr<T>(const HoldPtr<T> &p) : ptr_(p.ptr_) {}
+    HoldPtr<T>(HoldPtr<T> &&p) : ptr_(p.ptr_) {
+        p.ptr_ = nullptr;
+    }
+    ~HoldPtr<T>() = default;
```
Comment (critical27) [inline src/common/base/HoldPtr.h:37]: Consider add copy assignment and move assignment. Or maybe declare as deleted.

**Review by critical27 — COMMENTED** [review]
Good job

File: src/common/base/HoldPtr.h (hunk 2 above)
Comment (Shylock-Hg) [reply src/common/base/HoldPtr.h:37]: They are supported.

File: src/common/base/HoldPtr.h (hunk 1 above)
Comment (Shylock-Hg) [reply src/common/base/HoldPtr.h]: Fixed.

**Pull request closed without merging**